quickwit-indexing/src/actors/ingest_api_garbage_collector.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use quickwit_actors::{Actor, ActorContext, ActorExitStatus, Handler, Mailbox};
+use quickwit_ingest_api::IngestApiService;
+use quickwit_metastore::Metastore;
+use quickwit_proto::ingest_api::{DropQueueRequest, ListQueuesRequest};
+use tracing::info;
+
+use super::IndexingService;
+use crate::actors::indexing_service::INGEST_API_SOURCE_ID;
+use crate::models::ShutdownPipeline;
+
+const RUN_INTERVAL: Duration = Duration::from_secs(60 * 60); // 1 hour
+
+#[derive(Debug, Clone, Default)]
+pub struct IngestApiGarbageCollectorCounters {
+    /// The number of passes the garbage collector has performed.
+    pub num_passes: usize,
+    /// The number of deleted queues.
+    pub num_deleted_queues: usize,
+}
+
+#[derive(Debug)]
+struct Loop;
+
+/// An actor for deleting not needed ingest api queues.
+pub struct IngestApiGarbageCollector {
+    metastore: Arc<dyn Metastore>,
+    ingest_api_service: Mailbox<IngestApiService>,
+    indexing_service: Mailbox<IndexingService>,
+    counters: IngestApiGarbageCollectorCounters,
+}
+
+impl IngestApiGarbageCollector {
+    pub fn new(
+        metastore: Arc<dyn Metastore>,
+        ingest_api_service: Mailbox<IngestApiService>,
+        indexing_service: Mailbox<IndexingService>,
+    ) -> Self {
+        Self {
+            metastore,
+            ingest_api_service,
+            indexing_service,
+            counters: IngestApiGarbageCollectorCounters::default(),
+        }
+    }
+
+    async fn delete_queue(&self, queue_id: &str) -> Result<(), ActorExitStatus> {
+        // shutdown the pipeline if any
+        self.indexing_service
+            .ask_for_res(ShutdownPipeline {
+                index_id: queue_id.to_string(),
+                source_id: INGEST_API_SOURCE_ID.to_string(),
+            })
+            .await
+            .map_err(|error| anyhow::anyhow!(error))?;
```

Comment:
This actor is unsupervized. It means that if an error is emitted in a handler, the actor will terminated and will not be restarted again.

We may want to be careful about the meaning and the effect of returning an error here.

I suggest we just log an error if that happens.